VoodooI2C/VoodooI2C/VoodooI2CDevice/VoodooI2CDeviceNub.cpp

```diff
@@ -72,20 +71,120 @@ IOReturn VoodooI2CDeviceNub::enableInterrupt(int source) {
     }
 }
 
+IOReturn VoodooI2CDeviceNub::evaluateDSM(const char *uuid, UInt32 index, OSObject **result) {
+    IOReturn ret;
+    uuid_t guid;
+    uuid_parse(uuid, guid);
+
+    // convert to mixed-endian
+    *(reinterpret_cast<uint32_t *>(guid)) = OSSwapInt32(*(reinterpret_cast<uint32_t *>(guid)));
+    *(reinterpret_cast<uint16_t *>(guid) + 2) = OSSwapInt16(*(reinterpret_cast<uint16_t *>(guid) + 2));
+    *(reinterpret_cast<uint16_t *>(guid) + 3) = OSSwapInt16(*(reinterpret_cast<uint16_t *>(guid) + 3));
+
+    OSObject *params[] = {
+        OSData::withBytes(guid, 16),
+        OSNumber::withNumber(I2C_DSM_REVISION, 32),
+        OSNumber::withNumber(index, 32),
+        OSArray::withCapacity(1),
+    };
+
+    ret = acpi_device->evaluateObject("_DSM", result, params, 4);
+    if (ret != kIOReturnSuccess)
+        ret = acpi_device->evaluateObject("XDSM", result, params, 4);
+
+    params[0]->release();
+    params[1]->release();
+    params[2]->release();
+    params[3]->release();
+    return ret;
+}
+
+IOReturn VoodooI2CDeviceNub::getDeviceResourcesDSM(UInt32 index, OSObject **result) {
+    if (evaluateDSM(I2C_DSM_TP7G, DSM_SUPPORT_INDEX, result) != kIOReturnSuccess) {
+        IOLog("%s::%s Could not find suitable _DSM or XDSM method\n", controller_name, getName());
+        return kIOReturnNotFound;
+    }
+
+    OSData *data = OSDynamicCast(OSData, *result);
+
+    if (!data) {
+        IOLog("%s::%s Could not get valid return for available TP7G indexes\n", controller_name, getName());
+        return kIOReturnNotFound;
+    }
+
+    UInt8 availableIndex = *(reinterpret_cast<UInt8 const*>(data->getBytesNoCopy()));
+    data->release();
+
+    if (!(availableIndex & (index << 1))) {
+        IOLog("%s::%s TP7G index 0x%x is not supported\n", controller_name, getName(), availableIndex);
+        return kIOReturnUnsupportedMode;
+    }
+
+    return evaluateDSM(I2C_DSM_TP7G, index, result);
+}
+
+IOReturn VoodooI2CDeviceNub::getAlternativeInterrupt(VoodooI2CACPICRSParser* crs_parser) {
+    OSObject *result = nullptr;
+    if (getDeviceResourcesDSM(TP7G_GPIO_INDEX, &result) != kIOReturnSuccess) {
+        IOLog("%s::%s failed to retrieve interrupts from _DSM or XDSM method\n", controller_name, getName());
+        result->release();
```

Comment:
You probably want to use `OSSafeReleaseNULL` here.
As it's in an error-handling context, there's a high possibility of `result` being `nullptr`.

You might want to enforce the usage of this macro in other places too. It also saves a manual check for null, and prevents accidental undefined behaviors.